src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs

```diff
@@ -332,83 +353,98 @@ public override bool TryWrite(T item)
-                        return true;
-                    }
-                    else
-                    {
-                        // The channel is full, and we're in a dropping mode.
-                        // Drop either the oldest or the newest and write the new item.
-                        if (parent._mode == BoundedChannelFullMode.DropNewest)
+                        else if (count < parent._bufferedCapacity)
+                        {
+                            // There's room in the channel.  Since we're not transitioning from 0-to-1 and
+                            // since there's room, we can simply store the item and exit without having to
+                            // worry about blocked/waiting readers.
+                            parent._items.EnqueueTail(item);
+                            return true;
+                        }
+                        else if (parent._mode == BoundedChannelFullMode.Wait)
+                        {
+                            // The channel is full and we're in a wait mode.
+                            // Simply exit and let the caller know we didn't write the data.
+                            return false;
+                        }
+                        else if (parent._mode == BoundedChannelFullMode.DropWrite)
                         {
-                            parent._items.DequeueTail();
+                            // The channel is full.  Just ignore the item being added
+                            // but say we added it.
+                            dropitem = DropItem.Create(item);
+                            return true;
                         }
                         else
                         {
-                            parent._items.DequeueHead();
+                            // The channel is full, and we're in a dropping mode.
+                            // Drop either the oldest or the newest and write the new item.
+                            if (parent._mode == BoundedChannelFullMode.DropNewest)
+                            {
+                                var droppedItem = parent._items.DequeueTail();
+                                dropitem = DropItem.Create(droppedItem);
+                            }
+                            else
+                            {
+                                var droppedItem = parent._items.DequeueHead();
+                                dropitem = DropItem.Create(droppedItem);
+                            }
+                            parent._items.EnqueueTail(item);
+                            return true;
                         }
-                        parent._items.EnqueueTail(item);
-                        return true;
+                    }
+                }
+                finally
+                {
+                    if (dropitem.HasValue)
+                    {
+                        _parent._itemDropped?.Invoke(dropitem.Item);
                     }
                 }
```

Comment:
I'm not in love with the idea of adding the extra try/finally, as it adds complexity and overhead.  A couple of alternatives:
1. Change the lock to be a try/finally, and just do the invocation as part of that try/finally after the call to Monitor.Exit.
2. Change the lock to be a try/finally, but also Exit the monitor prior to each return statement, then invoking the callback between the exit and the return, e.g.
```C#
bool releaseLock = false;
try
{
    Monitor.Enter(ref releaseLock);
    ...
    T droppedItem = parent._mode == BoundedChannelFullMode.DropNewest ?
        parent._items.DequeueTail() :
        parent._items.DequeueHead();
    parent._items.EnqueueTail(item);
    Monitor.Exit(parent.SyncObj);
    releaseLock = false;
    _parent._itemDropped?.Invoke(droppedItem);
    return true;
    ...
}
finally
{
    if (releaseLock)
    {
        Monitor.Exit(parent.SyncObj);
    }
}

Reply:
Thanks.
Refactored according to suggestion.